airflow/jobs/job.py

```diff
@@ -222,34 +222,23 @@ def heartbeat(self, only_if_necessary: bool = False) -> None:
             # We didn't manage to heartbeat, so make sure that the timestamp isn't updated
             self.latest_heartbeat = previous_heartbeat
 
-    def run(self) -> int | None:
-        """Starts the job."""
+    @provide_session
+    def prepare_for_execution(self, session: Session = NEW_SESSION):
```

Comment:
Since the method and many others in the `run_job` chain of calls are all internal-only and always share one single session, the `@provide_session` decorator is unnecessary and only open possibilities for mis-uses. This would be better IMO

```suggestion
    def prepare_for_execution(self, *, session: Session) -> None:
```

Also it’d be a good idea to add `:meta private:` to all of these functions.

Reply:
Good ideas. I am updating this part now and even adding some sequence diagrams to show how things "will" work when we implement AIP-44